pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,347 @@
+                    return progress.upload_failure, uploaded_count
+
+            # We didn't find single file with the matching hexdigest.
+            # Report it as missing but keep uploading other files.
+            return progress.upload_success, uploaded_count
+
+        def _result_cb(*, map_in, map_out):
+            # progress callback in 'main' thread
+            progress_callback, count = map_out
+            result_data["files"] += count
+            progress_callback(map_in)  # hexdigest
+            return True
+
+        sorted_todo_hexdigests = sorted(
+            todo_hexdigests, key=lambda hexdigest: -snapshotter.hexdigest_to_snapshotfiles[hexdigest][0].file_size
+        )
+        if not parallel_map_to(
+            fun=_upload_files_in_thread, iterable=sorted_todo_hexdigests, result_callback=_result_cb, n=self.parallel
+        ):
+            progress.add_fail()
+
+        progress.done()
+
+        if progress.finished_successfully:
+            self.log.info("All delta file were submitted for uploading")
+        else:
+            raise BackupFailure("Error while submitting delta backup files for uploading")
+
+        return result_data["files"]
+
+    def _delta_upload(self, snapshot_result: SnapshotResult, snapshotter: Snapshotter, start_time_utc):
+        callback_queue = Queue()
+
+        total_stored_size = 0
+        total_size = 0
+
+        # Determine which digests already exist and which need to be uploaded, also restore the backup size of re-used
+        # files from manifests
+        snapshot_hashes = set(snapshot_result.hashes)
+        uploaded_hashes = set()
+        all_existing_files = self._list_existing_files()
+        for snapshot_file in snapshot_result.state.files:
+            if snapshot_file.hexdigest in all_existing_files:
+                snapshot_file_from_manifest = all_existing_files[snapshot_file.hexdigest]
+                total_stored_size += snapshot_file_from_manifest.stored_file_size
+                total_size += snapshot_file_from_manifest.file_size
+                uploaded_hashes.add(
+                    SnapshotHash(
+                        hexdigest=snapshot_file_from_manifest.hexdigest, size=snapshot_file_from_manifest.file_size
+                    )
+                )
+
+        todo = snapshot_hashes.difference(uploaded_hashes)
+        todo_count = len(todo)
+
+        self.log.info(
+            "Submitting delta hashes to upload: %r, total hashes in the snapshot: %r", todo_count, len(snapshot_hashes)
+        )
+
+        upload_files_count = self._submit_uploading(callback_queue=callback_queue, todo=todo, snapshotter=snapshotter)
+
+        processed_files_count = 0
+        process_futures = []
+        upload_size = 0
+        upload_stored_size = 0
+
+        self.log.info("Waiting for the upload of delta backup files (%r) to finish", upload_files_count)
+
+        def _post_process_delta_file(upload_result):
```

Comment:
Looks like result could also be failure but I don't see any handling for that. I suppose if uploading any file failed the entire backup should be considered to have failed (or then we need to retry the failed file until it succeeds). If single failure means the entire backup failed checking results only at the very end could be quite wasteful as we could've upload a lot of files totally in vain. Also if the backup is considered to have failed we should presumably delete any files that didn't yet exist.

Reply:
I've changed it to the same logic we have for usual basebackups (with retries), moved to transfer.py